File: app/src/test/java/dev/bee/kanjianki/update/AutoUpdateWorkerTest.java
package dev.bee.kanjianki.update;

import androidx.work.ListenableWorker;

import org.junit.Test;

import java.util.concurrent.atomic.AtomicInteger;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

public final class AutoUpdateWorkerTest {
    @Test
    public void autoUpdateDecisionSkipsDisabledAndPendingUpdates() {
        AtomicInteger checks = new AtomicInteger();

        ListenableWorker.Result disabled = AutoUpdateWorker.runAutoUpdate(
                false,
                false,
                () -> {
                    checks.incrementAndGet();
                    return result(false);
                }
        );
        ListenableWorker.Result pending = AutoUpdateWorker.runAutoUpdate(
                true,
                true,
                () -> {
                    checks.incrementAndGet();
                    return result(false);
                }
        );

        assertSuccess(disabled);
        assertSuccess(pending);
        assertEquals(0, checks.get());
    }

    @Test
    public void autoUpdateDecisionMapsCheckerResultToWorkerOutcome() {
        assertSuccess(AutoUpdateWorker.runAutoUpdate(true, false, () -> result(false)));
        assertRetry(AutoUpdateWorker.runAutoUpdate(true, false, () -> result(true)));
    }

    @Test
    public void automaticUpdateCheckerFactoryRunsInjectedAutomaticRunner() {
        AtomicInteger checks = new AtomicInteger();

        AutoUpdateWorker.UpdateChecker checker = AutoUpdateWorker.automaticUpdateCheckerFactory(context -> {
            checks.incrementAndGet();
            return result(false);
        }).create(null);

        GitHubUpdater.UpdateResult result = checker.check();

        assertEquals(1, checks.get());
        assertEquals("done", result.message);
    }

    private static GitHubUpdater.UpdateResult result(boolean retryable) {
        return new GitHubUpdater.UpdateResult(!retryable, retryable ? "retry later" : "done", null, false, retryable);
    }

    private static void assertSuccess(ListenableWorker.Result result) {
        assertTrue(result instanceof ListenableWorker.Result.Success);
    }

    private static void assertRetry(ListenableWorker.Result result) {
        assertTrue(result instanceof ListenableWorker.Result.Retry);
    }
}

File: app/src/test/java/dev/bee/kanjianki/update/AutoUpdateWorkerTest.kt
package dev.bee.kanjianki.update

import androidx.work.ListenableWorker
import dev.bee.kanjianki.updatecore.AutoUpdateRunPolicy
import java.util.concurrent.atomic.AtomicInteger
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

class AutoUpdateWorkerTest {
    @Test
    fun autoUpdateDecisionSkipsDisabledAndPendingUpdates() {
        val checks = AtomicInteger()

        val disabled = AutoUpdateWorker.runAutoUpdate(
            false,
            false,
        ) {
            checks.incrementAndGet()
            result(false)
        }
        val pending = AutoUpdateWorker.runAutoUpdate(
            true,
            true,
        ) {
            checks.incrementAndGet()
            result(false)
        }

        assertSuccess(disabled)
        assertSuccess(pending)
        assertEquals(0, checks.get())
    }

    @Test
    fun autoUpdateDecisionMapsCheckerResultToWorkerOutcome() {
        assertSuccess(AutoUpdateWorker.runAutoUpdate(true, false) { result(false) })
        assertRetry(AutoUpdateWorker.runAutoUpdate(true, false) { result(true) })
        assertSuccess(AutoUpdateWorker.workerResult(AutoUpdateRunPolicy.WorkerOutcome.SUCCESS))
        assertRetry(AutoUpdateWorker.workerResult(AutoUpdateRunPolicy.WorkerOutcome.RETRY))
    }

    @Test
    fun automaticUpdateCheckerFactoryRunsInjectedAutomaticRunner() {
        val checks = AtomicInteger()

        val checker = AutoUpdateWorker.automaticUpdateCheckerFactory { _ ->
            checks.incrementAndGet()
            result(false)
        }.create(null)

        val result = checker.check()

        assertEquals(1, checks.get())
        assertEquals("done", result.message)
    }
}

private fun result(retryable: Boolean): GitHubUpdater.UpdateResult =
    GitHubUpdater.UpdateResult(!retryable, if (retryable) "retry later" else "done", null, false, retryable)

private fun assertSuccess(result: ListenableWorker.Result) {
    assertTrue(result is ListenableWorker.Result.Success)
}

private fun assertRetry(result: ListenableWorker.Result) {
    assertTrue(result is ListenableWorker.Result.Retry)
}
